### PR TITLE
fix flap in no_holes_when_write_suffix_failed

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix flap in test_storage_kafka/test.py::test_kafka_no_holes_when_write_suffix_failed (✕1)

Detailed description / Documentation draft:
https://clickhouse-test-reports.s3.yandex.net/15199/5b8630e11894266bc9d8a8841c712158a0866123/integration_tests_flaky_check_(asan).html#fail1 

It seems assignment took longer than 12 sec:
```
2020.10.07 23:32:39.485454 [ 20 ] {} <Debug> StorageKafka (kafka): Started streaming to 1 attached views
2020.10.07 23:32:39.487831 [ 20 ] {} <Trace> StorageKafka (kafka): Already subscribed to topics: []
2020.10.07 23:32:39.487946 [ 20 ] {} <Trace> StorageKafka (kafka): Already assigned to: [  ]
2020.10.07 23:32:42.500536 [ 20 ] {} <Trace> StorageKafka (kafka): Topics/partitions assigned: [ no_holes_when_write_suffix_failed[0:#] ]
2020.10.07 23:32:42.604607 [ 20 ] {} <Trace> StorageKafka (kafka): Polled batch of 20 messages. Offsets position: [ no_holes_when_write_suffix_failed[0:20] ]

2020.10.07 23:33:02.610366 [ 20 ] {} <Debug> test.view (bde3a7fe-983b-4ae7-86e2-2bd1f124a3f8) (Replicated OutputStream): Wrote block with ID 'all_13234067450316952167_1031822982405810513', 20 rows
2020.10.07 23:33:02.616476 [ 20 ] {} <Trace> test.view (bde3a7fe-983b-4ae7-86e2-2bd1f124a3f8): Renaming temporary part tmp_insert_all_1_1_0 to all_0_0_0.
2020.10.07 23:33:02.620250 [ 24 ] {} <Debug> test.view (ReplicatedMergeTreeQueue): Pulling 1 entries to queue: log-0000000000 - log-0000000000

2020.10.07 23:33:02.619621 [ 20 ] {} <Trace> StorageKafka (kafka): Polled offset 20 (topic: no_holes_when_write_suffix_failed, partition: 0)
2020.10.07 23:33:02.621509 [ 20 ] {} <Trace> StorageKafka (kafka): Committed offset 20 (topic: no_holes_when_write_suffix_failed, partition: 0)

2020.10.07 23:33:02.621744 [ 20 ] {} <Debug> StorageKafka (kafka): Started streaming to 1 attached views
2020.10.07 23:33:02.624161 [ 20 ] {} <Trace> StorageKafka (kafka): Already subscribed to topics: [no_holes_when_write_suffix_failed]
2020.10.07 23:33:02.624224 [ 20 ] {} <Trace> StorageKafka (kafka): Already assigned to: [ no_holes_when_write_suffix_failed[0:#] ]
2020.10.07 23:33:03.125022 [ 20 ] {} <Trace> StorageKafka (kafka): Polled batch of 2 messages. Offsets position: [ no_holes_when_write_suffix_failed[0:22] ]
2020.10.07 23:33:03.625300 [ 20 ] {} <Trace> StorageKafka (kafka): Stalled
2020.10.07 23:33:04.125436 [ 20 ] {} <Trace> StorageKafka (kafka): Stalled
2020.10.07 23:33:04.625579 [ 20 ] {} <Trace> StorageKafka (kafka): Stalled
```

### CODE REVIEW

**Comment by filimonov** [comment]
Hm... Strange, lot of unrelated things blew up, timing looks ok:
```
2020.10.08 18:14:36.836185 [ 37 ] {2c98244f-06a5-4c1f-b421-cd105f954725} <Debug> executeQuery: (from 172.18.0.1:48722) CREATE TABLE test.view (key UInt64, value String) ENGINE = ReplicatedMergeTree('/clickhouse/kafkatest/tables/no_holes_when_write_suffix_failed', 'node1') ORDER BY key; 

2020.10.08 18:14:36.572675 [ 259 ] {0d4b9294-b3ab-4111-a1f3-8099d6f1d8b9} <Trace> StorageKafka (kafka): Committed offset 1 (topic: no_holes_when_write_suffix_failed, partition: 0)

2020.10.08 18:14:37.062885 [ 25 ] {} <Debug> StorageKafka (kafka): Started streaming to 1 attached views
2020.10.08 18:14:37.065252 [ 25 ] {} <Trace> StorageKafka (kafka): Already subscribed to topics: [no_holes_when_write_suffix_failed]
2020.10.08 18:14:37.065310 [ 25 ] {} <Trace> StorageKafka (kafka): Already assigned to: [ no_holes_when_write_suffix_failed[0:#] ]
2020.10.08 18:14:37.065842 [ 25 ] {} <Trace> StorageKafka (kafka): Polled batch of 20 messages. Offsets position: [ no_holes_when_write_suffix_failed[0:21] ]

2020.10.08 18:14:37.068370 [ 25 ] {} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2020.10.08 18:14:57.070143 [ 25 ] {} <Debug> DiskLocal: Reserving 1.00 MiB on disk `default`, having unreserved 348.44 GiB.
2020.10.08 18:14:57.071631 [ 25 ] {} <Debug> test.view (9422974c-eb8e-4930-8ada-4d0f0e32120c) (Replicated OutputStream): Wrote block with ID 'all_13234067450316952167_1031822982405810513', 20 rows
2020.10.08 18:14:57.077626 [ 25 ] {} <Trace> test.view (9422974c-eb8e-4930-8ada-4d0f0e32120c): Renaming temporary part tmp_insert_all_1_1_0 to all_0_0_0.

2020.10.08 18:14:57.080504 [ 25 ] {} <Trace> StorageKafka (kafka): Polled offset 21 (topic: no_holes_when_write_suffix_failed, partition: 0)
2020.10.08 18:14:57.082124 [ 25 ] {} <Trace> StorageKafka (kafka): Committed offset 21 (topic: no_holes_when_write_suffix_failed, partition: 0)
```

File: tests/integration/test_storage_kafka/test.py [hunk 1]
```diff
@@ -1833,12 +1840,13 @@ def test_kafka_no_holes_when_write_suffix_failed(kafka_cluster):
...
+
     # the tricky part here is that disconnect should happen after write prefix, but before write suffix
     # so i use sleepEachRow
     with PartitionManager() as pm:
-        time.sleep(12)
+        time.sleep(5)
         pm.drop_instance_zk_connections(instance)
-        time.sleep(20)
+        time.sleep(21)
         pm.heal_all
```
Comment (azat) [inline tests/integration/test_storage_kafka/test.py:1850]:
```suggestion
        pm.heal_all()
```

**Comment by filimonov** [comment]
Something strange:
```
2020.10.09 13:47:24.263212 Before drop_instance_zk_connections
2020.10.09 13:51:54.287981 After drop_instance_zk_connections
```

4 minutes!
https://github.com/ClickHouse/ClickHouse/blob/f800844da281c6c79f9a721dfffa887a1db0fffd/tests/integration/test_storage_kafka/test.py#L1848-L1850

**Comment by filimonov** [comment]
Now it's on top of #15786

**Comment by alesapin** [comment]
Seems like `test_storage_kafka/test.py::test_kafka_produce_key_timestamp (✕5)` indeed flaky?

**Comment by filimonov** [comment]
Yep. San't reproduce that locally too, but from from logs it seems caused by some kind of race between producer and consumer.
Since rdkafka 1.5 consumer don't create topic automatically and for some reason rdkafka on the next attempt start reading from offset 2, and get 4 messages instead of 5.

```
2020.10.09 20:31:08.152633 [ 354 ] {} <Debug> StorageKafka (kafka): [rdk:TOPICERR] [thrd:main]: Subscribed topic not available: insert3: Broker: Unknown topic or partition
... 
2020.10.09 20:31:12.160404 [ 354 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: insert3 [0]: offset reset (at offset INVALID) to BEGINNING: no previously committed offset available: Local: No offset stored
2020.10.09 20:31:12.160436 [ 354 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: insert3 [0]: backoff: (re)starting offset query timer for offset BEGINNING
...
2020.10.09 20:31:12.261674 [ 354 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: Offset BEGINNING request for insert3 [0] returned offset 2 (2)
2020.10.09 20:31:12.261724 [ 354 ] {} <Debug> StorageKafka (kafka): [rdk:FETCH] [thrd:main]: Partition insert3 [0] start fetching at offset 2
2020.10.09 20:31:12.656422 [ 22 ] {} <Trace> StorageKafka (kafka): Polled batch of 2 messages. Offsets position: [ insert3[0:4] ]
2020.10.09 20:31:13.157760 [ 22 ] {} <Trace> StorageKafka (kafka): Polled batch of 1 messages. Offsets position: [ insert3[0:5] ]
2020.10.09 20:31:16.675840 [ 22 ] {} <Trace> StorageKafka (kafka): Polled offset 5 (topic: insert3, partition: 0)
```

I think explicit topic creation should help

**Comment by filimonov** [comment]
AST fuzzer - already reported: https://github.com/ClickHouse/ClickHouse/issues/14974

**Review by alexey-milovidov — COMMENTED** [review]
It's all related to timing, how it can work reliably?

**Comment by filimonov** [comment]
It's not 100% reliable, but time windows are quite wide (dozens of seconds) and those tests normally work good enough and don't flap (they exis in CI for many months already). 

And there are lot of timing invariants in Kafka anyway (for example if you will not do poll at least once per 60 sec you will be kicked off by broker), so if you can't give some guarantees about timing you can't test Kafka client at all.

**Comment by alexey-milovidov** [comment]
> normally work good enough and don't flap

:(

**Comment by filimonov** [comment]
> > normally work good enough and don't flap
> 
> :(

It's not possible to make all integration tests 100% reliable (unless you mock all the external parts, which has no sense at all as it's too much work). 

Under normal condition those tests should not flap, in some special conditions (thread fuzzer, etc) those time-related tests should be just tagged special way and skipped.

**Pull request merged**